### PR TITLE
Refactor Security Class

### PR DESCRIPTION
**Pull request by mostafakhudair**
Creates new Config\Security class and make \CodeIgniter\Security\Security class depends on it

removes useless properties
rename some properties

rename some methods
```
CSRFVerify() to verify()
getCSRFHash() to getHash()
getCSRFTokenName() to getTokenName()
setHash() to generateHash()
```

add new public methods:
```
getHeaderName()
getCookieName()
isExpired()
shouldRedirect()
```

edits changes in user guide

- [x] Securely signed commits
- [x] Component(s) with PHPdocs
- [x] User guide updated
- [x] Conforms to style guide

Closes #3836

### CODE REVIEW

File: system/Config/Services.php [hunk 1]
```diff
@@ -677,7 +677,7 @@ public static function security(App $config = null, bool $getShared = true)
 			return static::getSharedInstance('security', $config);
 		}
 
-		$config = $config ?? config('App');
+		$config = $config ?? config('App') ?? config('Security');
```
Comment (paulbalandan) [inline system/Config/Services.php:680]: I think the check for Security config should come first before App.
```suggestion
		$config = $config ?? config('Security')  ?? config('App');
```

File: system/Security/Security.php [hunk 2]
```diff
@@ -411,4 +376,89 @@ public function sanitizeFilename(string $str, bool $relativePath = false): strin
...
+	protected function generateHash(): string
+	{
+		if (is_null($this->hash))
+		{
+			// If the cookie exists we will use its value.
+			// We don't necessarily want to regenerate it with
+			// each page load since a page could contain embedded
+			// sub-pages causing this feature to fail
+			if (isset($_COOKIE[$this->cookieName]) && is_string($_COOKIE[$this->cookieName]) && preg_match('#^[0-9a-f]{32}$#iS', $_COOKIE[$this->cookieName]) === 1
+			)
```
Comment (paulbalandan) [inline system/Security/Security.php:394]: Can we break this in viewable line breaks?

**Review by paulbalandan — COMMENTED** [review]
The refactor is good to me but we need the review of others since this is a major component of the framework. Some notes:

- Fix the user guide syntax check
- In the main Security class, you cannot rename or remove the public and protected methods as it would mean a BC break for people using those. The deprecations and detours for the renamed Security methods are good though.
- See other comments below.

**Comment by MGatner** [comment]
This looks great! I'm a fan of splitting these out, you've been doing a great job with it. I'd like to give this one more look when I'm not on mobile (or have someone else's review) but let's plan to include this in 4.1 so these deprecations can get entered.

**Comment by MGatner** [comment]
Is there any merit to merging this and ContentSecurityPolicy config? I'm not terribly knowledgeable of the various security components but as a novice having all "security" under one Config seems logical?

**Comment by mostafakhudair** [comment]
> Is there any merit to merging this and ContentSecurityPolicy config? I'm not terribly knowledgeable of the various security components but as a novice having all "security" under one Config seems logical?

While ContentSecurityPolicy config serve HTTP\ContentSecurityPolicy class, and Security config serve Security\Security class, I think there is no point in merging them together at least for now.

**Comment by lonnieezell** [comment]
I think this looks good also. @MGatner are we good to merge this into 4.0.5 or should this be a 4.1 change? It looks like everything is deprecated and BC safe.

**Comment by MGatner** [comment]
@lonnieezell I think this is fine for 4.0.5. Since we decided to support PHP 8 in 4.0.5 the only real rubric for 4.1 is support for PHP 7.2.

**Pull request merged**